### PR TITLE
Enable travis to publish to npm.

### PR DESCRIPTION
**Pull request by matthew-andrews**
Perhaps a little overzealous of me to rewrite release process but I think less is more :smile: .  Very happy to raise a simpler PR that **just** does the `npm publish` parts if you'd prefer to keep the existing process.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
   "name": "fetch",
-  "version": "0.4.0",
   "main": "fetch.js",
   "private": true,
   "devDependencies": {
     "bower": "1.3.8",
     "jshint": "2.5.2",
-    "node-qunit-phantomjs": "0.2.2"
+    "node-qunit-phantomjs": "0.2.2",
+    "npm-prepublish": "1.0.1"
```
Comment (matthew-andrews) [inline package.json]: out of curiosity why is it you lock down your versions of dependencies?

**Comment by matthew-andrews** [comment]
Also to publish under the “fetch” name [as offered by](https://github.com/github/fetch/issues/9#issuecomment-59704339) @andris9 you'll need to `git tag v2.0.0` (we could +1 to the major version just for npm, but that's likely to get super confusing…

File: package.json (hunk 1 above)
Comment (josh) [inline package.json]: So nothing randomly breaks without a deliberate change.

File: MAINTAINING.md [hunk 2]
```diff
@@ -6,24 +6,14 @@ This project follows [semver](http://semver.org/). So if you are making a bug
...
-3. Make a single commit with the description as "Fetch 1.x.x".
-4. Finally, tag the commit with `v1.x.x`.
+1. Tag the commit with `v1.x.x`.
 
 ```
 $ git pull
-$ vim bower.json
-$ vim package.json
-$ git add bower.json package.json
-$ git commit -m "Fetch 1.x.x"
```
Comment (josh) [inline MAINTAINING.md]: I would still like to tag an empty commit in that case.

File: .travis.yml [hunk 3]
```diff
@@ -1,2 +1,13 @@
 language: node_js
 script: make test
+before_deploy:
+- npm-prepublish --verbose
+deploy:
+  provider: npm
+  email: "replace-with-email"
+  api_key:
+    secure: "replace-with-encrypted-api-key"
```
Comment (josh) [inline .travis.yml]: Can we publish this on someone else's account? I don't really have any of this setup anymore.

File: MAINTAINING.md (hunk 2 above)
Comment (matthew-andrews) [inline MAINTAINING.md]: I'll update.

File: .travis.yml (hunk 3 above)
Comment (matthew-andrews) [inline .travis.yml]: I think it makes sense for someone who is a collaborator to do this @dgraham would you mind?

File: MAINTAINING.md (hunk 2 above)
Comment (josh) [inline MAINTAINING.md]: Cool, basically

```
$ git commit -allow-empty -m "Fetch 1.x.x"
```

Comment (matthew-andrews) [inline MAINTAINING.md]: error: did you mean `--allow-empty` (with two dashes ?)

File: .travis.yml (hunk 3 above)
Comment (matthew-andrews) [inline .travis.yml]: Otherwise I can but if it breaks I'd be limited in what I can do :-/

File: package.json (hunk 1 above)
Comment (matthew-andrews) [inline package.json]: Ah yeah.  shrinkwrap doesn't happen for devDeps.  Makes sense.

**Comment by matthew-andrews** [comment]
What do you want to do about the npm registry conflict?  Bump to v2.x.x or rename to something like `whatwg-fetch` ?

**Comment by matthew-andrews** [comment]
Oh wait, I knew this would be too good to be true.

Removing `version` from `package.json` this breaks installing via `npm install github/fetch` because that requires `version` to be defined.

**Comment by josh** [comment]
> What do you want to do about the npm registry conflict? Bump to v2.x.x or rename to something like whatwg-fetch ?

Yeah, I'd rather not change our versioning policy to fit it. npm can just be called `whatwg-fetch`. And still use `fetch` on bower.

**Comment by matthew-andrews** [comment]
OK, I've put in my encrypted deploy key for now.  Final thing… I found a workaround to the version-in-package.json problem:-

Would you prefer to leave the real version number in `package.json` (as it is now) and have to update that on every release or would you rather set the version in `package.json` to be `0.0.0` [(described here)](https://github.com/matthew-andrews/npm-prepublish#warning)? (one less thing to update)

**Comment by matthew-andrews** [comment]
[If you're going to update this manually then don't need npm-prepublish at all and can ditch that dependency & complexity]

**Comment by josh** [comment]
> Would you prefer to leave the real version number in package.json (as it is now) and have to update that on every release 

I don't mind bumping the version. Its cool with me.

File: MAINTAINING.md [hunk 4]
```diff
@@ -10,18 +10,15 @@ minor version "1.x.x" bump is in order.
 
 To prepare the release commit:
 
-1. Edit the [bower.json](https://github.com/github/fetch/blob/master/bower.json)
+1. Edit the npm [package.json](https://github.com/github/fetch/blob/master/package.json)
 `version` value.
-2. Change the npm [package.json](https://github.com/github/fetch/blob/master/package.json)
-`version` value to match.
-3. Make a single commit with the description as "Fetch 1.x.x".
-4. Finally, tag the commit with `v1.x.x`.
```
Comment (josh) [inline MAINTAINING.md]: I'd be fine leaving everything the same here. I can just update both spots.

File: package.json (hunk 1 above)
Comment (josh) [inline package.json]: So we do or do not need npm-prepublish if I do all the version bumps by hand?

File: package.json [hunk 5]
```diff
@@ -1,11 +1,12 @@
 {
-  "name": "fetch",
+  "name": "whatwg-fetch",
   "version": "0.4.0",
   "main": "fetch.js",
   "private": true,
```
Comment (josh) [inline package.json]: Probably need to remove private.

**Comment by josh** [comment]
@matthew-andrews thanks for getting this going! I think we're close to trying it out!

File: package.json (hunk 1 above)
Comment (matthew-andrews) [inline package.json]: Don't need at all — funny, this is the thing that bothers me most about npm :smile:.  Will simplify massively.

**Comment by matthew-andrews** [comment]
ping @josh should be good to go now.

**Comment by matthew-andrews** [comment]
[sorry for the massive number of commits — if you like I can close this and tidy up the history?]

**Comment by josh** [comment]
lets do it

**Comment by josh** [comment]
I'm going to prep a release shorty.

**Comment by matthew-andrews** [comment]
hm.  it didn't work :-/.  let me investigate.

**Comment by josh** [comment]
```
Fetching: dpl-1.7.9.gem (100%)
Successfully installed dpl-1.7.9
1 gem installed
73.73s
dpl.1
Installing deploy dependencies
dpl.2
Preparing deploy
Authenticated with email matt@mattandre.ws
/usr/lib/git-core/git-stash: 186: /usr/lib/git-core/git-stash: cannot create /home/travis/build/github/fetch/.git/logs/refs/stash: Directory nonexistent
dpl.3
Deploying application
npm ERR! publish Failed PUT 500
npm ERR! Error: pound_internal_server_error : whatwg-fetch
npm ERR!     at RegClient.<anonymous> (/home/travis/.nvm/v0.10.35/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:308:14)
npm ERR!     at Request._callback (/home/travis/.nvm/v0.10.35/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:246:65)
npm ERR!     at Request.self.callback (/home/travis/.nvm/v0.10.35/lib/node_modules/npm/node_modules/request/request.js:236:22)
npm ERR!     at Request.emit (events.js:98:17)
npm ERR!     at Request.<anonymous> (/home/travis/.nvm/v0.10.35/lib/node_modules/npm/node_modules/request/request.js:1142:14)
npm ERR!     at Request.emit (events.js:117:20)
npm ERR!     at IncomingMessage.<anonymous> (/home/travis/.nvm/v0.10.35/lib/node_modules/npm/node_modules/request/request.js:1096:12)
npm ERR!     at IncomingMessage.emit (events.js:117:20)
npm ERR!     at _stream_readable.js:944:16
npm ERR!     at process._tickCallback (node.js:442:13)
npm ERR! If you need help, you may report this *entire* log,
npm ERR! including the npm and node versions, at:
npm ERR!     <http://github.com/npm/npm/issues>
npm ERR! System Linux 2.6.32-042stab090.5
npm ERR! command "/home/travis/.nvm/v0.10.35/bin/node" "/home/travis/.nvm/v0.10.35/bin/npm" "publish"
npm ERR! cwd /home/travis/build/github/fetch
npm ERR! node -v v0.10.35
npm ERR! npm -v 1.4.28
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/travis/build/github/fetch/npm-debug.log
npm ERR! not ok code 0
No stash found.
```

**Comment by josh** [comment]
No idea what the "stash" it

**Comment by matthew-andrews** [comment]
not too worried about that — seems to be something funky in travis — happens on working builds:-
https://travis-ci.org/matthew-andrews/npm-prepublish/jobs/46489270

suspect this is a bug.  i *think* (although i can't find the build now) this is just that travis won't let you publish the *first* publish of a npm module via its deployment mechanism…  going to verify if that is true with another module i'm working on right now…

**Comment by matthew-andrews** [comment]
argh.  i suspect the rename will have broken things for some consumers too.

anyone who was using `npm install github/fetch` was tracking master and so got the new name the moment we hit merge…  and then if they were using node or browserify they'll need to update their requires from `require('fetch')` to `require('whatwg-fetch')`

:disappointed:

**Comment by matthew-andrews** [comment]
hmm. no, it's not that. worked perfectly for me:-
https://travis-ci.org/matthew-andrews/fetchres/builds/46795881

**Comment by matthew-andrews** [comment]
This is my best guess https://github.com/github/fetch/pull/63/files :/

[ partially based on this https://github.com/npm/npm/issues/6669 ]

**Comment by domenic** [comment]
> anyone who was using npm install github/fetch was tracking master

Well that would have been very silly of them.

**Comment by matthew-andrews** [comment]
@domenic that's what the `fetch` docs used to advocate…

if we're playing the who's sillier than who game, i think it's sillier that npm doesn't provide a way of being able to install based on semver patterns via git urls, like bower can :P

**Comment by matthew-andrews** [comment]
also what's the point of npm providing an api like `npm install org/repo` if it's silly to use it…?

**Comment by matthew-andrews** [comment]
@josh just to keep the tag history clean, i think i'll manually publish `0.5.0` and i'll keep an eye on the new versions and make sure the future ones happen automatically

**Comment by mislav** [comment]
@matthew-andrews I've just pushed tag v0.10.0 but npm deploy seems to be failing possibly due to API keys having changed format? https://travis-ci.org/github/fetch/jobs/84903954

Could you help me out here?

```
Deploying application
NPM API key format changed recently. If your deployment fails, check your API key in ~/.npmrc.
http://docs.travis-ci.com/user/deployment/npm/
~/.npmrc size: 49
npm ERR! publish Failed PUT 500
npm ERR! Error: unexpected status code: 401 : whatwg-fetch
```

**Comment by matthew-andrews** [comment]
Yep that's right…  Do you want me to update the key or would you prefer to transfer this to your npm account?

**Comment by mislav** [comment]
Yeah I guess I'll just switch over to my credentials so we don't have to bug you each time this changes. Thanks!

**Comment by mislav** [comment]
@matthew-andrews So I'm [@mislav](https://www.npmjs.com/~mislav) on npm. How do we transfer your package?

**Comment by matthew-andrews** [comment]
@mislav you should be able to publish now :smile: 
```
npm owner add mislav whatwg-fetch
+ mislav (whatwg-fetch)
```

**Comment by mislav** [comment]
Thanks! That worked

**Pull request merged**